src/rust/bitbox02-rust/src/hww/api/bluetooth.rs

```diff
@@ -0,0 +1,273 @@
+        crate::hww::next_request(pb::response::Response::Bluetooth(pb::BluetoothResponse {
+            response: Some(response),
+        }))
+        .await?;
+    match request {
+        pb::request::Request::Bluetooth(pb::BluetoothRequest {
+            request: Some(request),
+        }) => Ok(request),
+        _ => Err(Error::InvalidState),
+    }
+}
+
+async fn get_fw_chunk_from_host(offset: u32, length: u32) -> Result<Vec<u8>, Error> {
+    let request = next_request(Response::RequestChunk(pb::BluetoothRequestChunkResponse {
+        offset,
+        length,
+    }))
+    .await?;
+    match request {
+        Request::Chunk(pb::BluetoothChunkRequest { data }) => Ok(data),
+        _ => Err(Error::InvalidInput),
+    }
+}
+
+trait Funcs {
+    async fn get_fw_chunk(&mut self, offset: u32, length: u32) -> Result<Vec<u8>, Error>;
+}
+
+async fn _process_upgrade(
+    funcs: &mut impl Funcs,
+    request: &pb::BluetoothUpgradeInitRequest,
+    allowed_hash: &[u8; 32],
+) -> Result<Response, Error> {
+    if request.firmware_length == 0 || request.firmware_length > spi_mem::BLE_FIRMWARE_MAX_SIZE {
+        return Err(Error::InvalidInput);
+    }
+
+    let mut ble_metadata = memory::get_ble_metadata();
+
+    // We work on the inactive firmware memory area.
+    let inactive_index: u8 = if ble_metadata.active_index == 0 { 1 } else { 0 };
+    let inactive_ble_fw_address = if inactive_index == 0 {
+        spi_mem::BLE_FIRMWARE_1_ADDR
+    } else {
+        spi_mem::BLE_FIRMWARE_2_ADDR
+    };
+
+    let mut firmware_hasher = Sha256::new();
+    let mut firmware_checksum = 0u8;
+
+    // The host needs to send this many chunks.
+    let num_chunks = (request.firmware_length + SPI_ERASE_SIZE - 1) / SPI_ERASE_SIZE;
+
+    // Show progress
+    let mut progress_component = bitbox02::ui::progress_create("Upgrading...");
+    progress_component.screen_stack_push();
+
+    // Stream chunks from host.
+    for chunk_index in 0..num_chunks {
+        let chunk_offset = chunk_index * SPI_ERASE_SIZE;
+        let chunk_length = core::cmp::min(SPI_ERASE_SIZE, request.firmware_length - chunk_offset);
+        let chunk: Vec<u8> = funcs.get_fw_chunk(chunk_offset, chunk_length).await?;
+        if chunk.len() != chunk_length as usize {
+            return Err(Error::InvalidInput);
+        }
+        firmware_hasher.update(&chunk);
+        for &byte in &chunk {
+            firmware_checksum ^= byte;
+        }
+
+        spi_mem::write(inactive_ble_fw_address + chunk_offset, &chunk)
+            .map_err(|_| Error::Memory)?;
+
+        // Update progress.
+        bitbox02::ui::progress_set(
+            &mut progress_component,
+            (chunk_index + 1) as f32 / (num_chunks as f32),
+        );
+    }
+
+    drop(progress_component);
+
+    let firmware_hash: [u8; 32] = firmware_hasher.finalize().into();
+    if &firmware_hash != allowed_hash {
+        return Err(Error::InvalidInput);
+    }
+
+    // TODO: read back and verify FW bytes
```

Comment:
This is quite an important TODO?

Reply:
It is, but let me add it in a separate PR :pray: